astropy/utils/data.py

```diff
@@ -982,11 +981,12 @@ def _download_file_from_source(source_url, show_progress=True, timeout=None,
                                http_headers=None, ftp_tls=None):
     from astropy.utils.console import ProgressBarOrSpinner
 
-    if timeout == 0:
+    if not conf.allow_internet:
```

Comment:
Do we have to deprecate `timeout == 0` before removing it outright? 🤷

Reply:
I cannot quite think straight whether it would ever be a problem, but I guess that if people have packages that depend on astropy and test with different versions, yet get their configuration from astropy, it would be. So, perhaps one should deprecate? But then that causes an error if the astropy version is new. So, maybe just check for `timeout == 0` and remove it the next time we look at this, for 5.0 or so (by which time anybody can change their code safely). And after all, it is not like downloading with timeout=0 is going to work...

Reply:
So you mean doing `if timeout == 0 or not conf.allow_internet` *without* deprecation?

Reply:
Yes. As I think with a deprecation warning, people will have to have astropy-version specific stuff in their tests, which seems rather much for tests. But perhaps there is a better solution?

Reply:
Yeah, I think a warning is overkill. `remote_timeout=0` was just merged. The only difference is that is going to LTS, and this is not. I agree that there is no harm to leave the old check in, albeit undocumented, as long as I can take it out in 5.0.

Reply:
No, actually, I take it back. I should at least keep the remote_timeout config comment. Otherwise it would be utterly surprising.

Reply:
I'd like to point out that `remote_timeout=0` is something you can easily pass to an individual `download_file` call without elaborate `with` statement config file jiggery-pokery. One can achieve a similar but not equivalent effect by passing `sources=[]` (this forbids non-network URLs too). So if you want to write a script for which one specific call doesn't touch the internet, this PR removes the ability to do that easily.

Reply:
I don't think we should worry too much about special cases, though I'd be quite happy to end up removing any check on time-out so that even zero just gets passed on. But I find code with lots of "look before you leap" stuff distracting, much preferring the let things fail as appropriate method...

Reply:
I am still trying to wrap my head around `sources` myself. I find that it is more an option for a "power user" and most of our users will just not touch it. But you bring up a good point, @aarchiba , we're have a problem of "too many ways to do the same thing" here. Not sure what is the best way forward.